crates/brainfuck_prover/src/components/memory/table.rs

```diff
@@ -18,229 +18,346 @@ use stwo_prover::{
+    ///
+    /// This method pushes a new [`MemoryTableRow`] onto the `table` vector.
+    fn add_row(&mut self, row: MemoryTableRow) {
+        self.table.push(row);
     }
 
-    /// Getter for the `clk` field.
-    pub const fn clk(&self) -> BaseField {
-        self.clk
-    }
+    /// Transforms the [`MemoryTable`] into [`super::super::TraceEval`], to be commited
+    /// when generating a STARK proof.
+    ///
+    /// The [`MemoryTable`] is transformed from an array of consecutive rows (one
+    /// element = one step of all registers) to an array of columns (one element = all steps of
+    /// one register). It is then evaluated on the circle domain.
+    ///
+    /// # Arguments
+    /// * memory - The [`MemoryTable`] containing the sorted and padded trace as an array of
+    ///   consecutive rows.
+    ///
+    /// # Returns
+    /// A tuple containing the evaluated trace and claim for STARK proof.
+    ///
+    /// # Errors
+    /// Returns [`TraceError::EmptyTrace`] if the table is empty.
+    pub fn trace_evaluation(&self) -> Result<(TraceEval, MemoryClaim), TraceError> {
+        let n_rows = self.table.len() as u32;
+        if n_rows == 0 {
+            return Err(TraceError::EmptyTrace);
+        }
+        let log_n_rows = n_rows.ilog2();
+        let log_size = log_n_rows + LOG_N_LANES;
+        let mut trace = vec![BaseColumn::zeros(1 << log_size); MemoryColumn::count()];
 
-    /// Getter for the `mp` field.
-    pub const fn mp(&self) -> BaseField {
-        self.mp
+        for (vec_row, row) in self.table.iter().enumerate().take(1 << log_n_rows) {
+            trace[MemoryColumn::Clk.index()].data[vec_row] = row.clk.into();
+            trace[MemoryColumn::Mp.index()].data[vec_row] = row.mp.into();
+            trace[MemoryColumn::Mv.index()].data[vec_row] = row.mv.into();
+            trace[MemoryColumn::D.index()].data[vec_row] = row.d.into();
+            trace[MemoryColumn::NextClk.index()].data[vec_row] = row.next_clk.into();
+            trace[MemoryColumn::NextMp.index()].data[vec_row] = row.next_mp.into();
+            trace[MemoryColumn::NextMv.index()].data[vec_row] = row.next_mv.into();
+            trace[MemoryColumn::NextD.index()].data[vec_row] = row.next_d.into();
+        }
+
+        let domain = CanonicCoset::new(log_size).circle_domain();
+        let trace = trace.into_iter().map(|col| CircleEvaluation::new(domain, col)).collect();
+
+        Ok((trace, MemoryClaim::new(log_size)))
     }
+}
 
-    /// Getter for the `mv` field.
-    pub const fn mv(&self) -> BaseField {
-        self.mv
+impl From<Vec<Registers>> for MemoryTable {
+    fn from(registers: Vec<Registers>) -> Self {
+        MemoryIntermediateTable::from(registers).into()
     }
+}
+
+// Separated from `Vec<Registers> for MemoryTable` to facilitate tests.
+// It is assumed that [`MemoryIntermediateTable`] is sorted and padded to the next power of two.
+impl From<MemoryIntermediateTable> for MemoryTable {
```

Comment:
If this is used only for tests we should put a test cfg flag on top of it no?

Reply:
It's used in `impl From<Vec<Registers>> for MemoryTable`

The `from(registers: Vec<Registers>) -> MemoryTable`  uses both `from(registers: Vec<Registers>) -> MemoryIntermediateTable` and `from(intermediate_table: MemoryIntermediateTable) -> MemoryTable` functions;

The comment was more like we could inline the logic of `from(intermediate_table: MemoryIntermediateTable) -> MemoryTable` within `from(registers: Vec<Registers>) -> MemoryTable` but it's useful to have it separated, so that the tests can be setup by directly declaring a `MemoryIntermediateTable` rather than a `Vec<Registers>`